Bu ödevde [ABD Eğitim Sistemi](https://www.kaggle.com/noriuk/us-education-datasets-unification-project/home)'ne  ait bir takım verileri inceleyeceksiniz. Öncelikle [veri kümesini](https://www.kaggle.com/noriuk/us-education-datasets-unification-project/home) Kaggle'dan indirin. Burada eğitim sistemine ait bir çok veri kümesi bulunmaktadır. Bu ödevde [**'states_all.csv'**](https://drive.google.com/open?id=1jbMrKbjMhQeG8xHUc8IFlOU1z-Eh3abc)  isimli veri kümesini kullanacaksınız. 

Bu ödevi tamamlamak için, aşağıdaki soruların çözümlerini içeren Jupyter notebook dosyasının Github linkini gönderin. Kafanıza takılan konuları mentörünüzle konuşabilir ya da ofis saatinde Slack üzerinden sorabilirsiniz.

1. Veri kümesinde yer alan değişkenlerinin tiplerini ve her bir değişkenin eksik (null) değer oranını bulun.

2. Verimizde yıl (year) sütunu olduğunu farketmişsinizdir. Şimdilik yıl verisini unutun ve her bir gözlemin aynı yıl içerisinde yapıldığını farz edin. Her bir değişken için eksik değerleri nasıl doldurabileceğinizi düşünün. Eksik değerleri bir değerle doldurmak  hangi değişkenler için anlamlı, hangileri için anlamsızdır?

3. Şimdi zaman faktörünü dikkate alma zamanı! 2. sorudaki cevabınızı tekrar gözden geçirin ve eksik verileri o yıl içerisinde gözlemlenen değerlere dayanarak doldurun. Örneğin, bir değeri ortalama değer ile doldurmak isterseniz, o yılın ortalamasını hesaplayın.

4. Bu sefer, eksik değerleri enterpolasyon yaparak doldurun.

5. İkinci, üçüncü ve dördüncü sorulardaki cevaplarınızı karşılaştırın. Anlamlı bir fark görebiliyor musunuz?

In [1]:
import pandas as pd
import numpy as np
import warnings

warnings.filterwarnings('ignore')

In [93]:
df=pd.read_csv('states_all.csv')
df.head(50)

PRIMARY_KEY  ... AVG_READING_8_SCORE
0                1992_ALABAMA  ...                 NaN
1                 1992_ALASKA  ...                 NaN
2                1992_ARIZONA  ...                 NaN
3               1992_ARKANSAS  ...                 NaN
4             1992_CALIFORNIA  ...                 NaN
5               1992_COLORADO  ...                 NaN
6            1992_CONNECTICUT  ...                 NaN
7               1992_DELAWARE  ...                 NaN
8   1992_DISTRICT_OF_COLUMBIA  ...                 NaN
9                1992_FLORIDA  ...                 NaN
10               1992_GEORGIA  ...                 NaN
11                1992_HAWAII  ...                 NaN
12                 1992_IDAHO  ...                 NaN
13              1992_ILLINOIS  ...                 NaN
14               1992_INDIANA  ...                 NaN
15                  1992_IOWA  ...                 NaN
16                1992_KANSAS  ...                 NaN
17              1992_KENTUCKY  ...                 NaN
18             1992_LOUISIANA  ...                 NaN
19                 1992_MAINE  ...                 NaN
20              1992_MARYLAND  ...                 NaN
21         1992_MASSACHUSETTS  ...                 NaN
22              1992_MICHIGAN  ...                 NaN
23             1992_MINNESOTA  ...                 NaN
24           1992_MISSISSIPPI  ...                 NaN
25              1992_MISSOURI  ...                 NaN
26               1992_MONTANA  ...                 NaN
27              1992_NEBRASKA  ...                 NaN
28                1992_NEVADA  ...                 NaN
29         1992_NEW_HAMPSHIRE  ...                 NaN
30            1992_NEW_JERSEY  ...                 NaN
31            1992_NEW_MEXICO  ...                 NaN
32              1992_NEW_YORK  ...                 NaN
33        1992_NORTH_CAROLINA  ...                 NaN
34          1992_NORTH_DAKOTA  ...                 NaN
35                  1992_OHIO  ...                 NaN
36              1992_OKLAHOMA  ...                 NaN
37                1992_OREGON  ...                 NaN
38          1992_PENNSYLVANIA  ...                 NaN
39          1992_RHODE_ISLAND  ...                 NaN
40        1992_SOUTH_CAROLINA  ...                 NaN
41          1992_SOUTH_DAKOTA  ...                 NaN
42             1992_TENNESSEE  ...                 NaN
43                 1992_TEXAS  ...                 NaN
44                  1992_UTAH  ...                 NaN
45               1992_VERMONT  ...                 NaN
46              1992_VIRGINIA  ...                 NaN
47            1992_WASHINGTON  ...                 NaN
48         1992_WEST_VIRGINIA  ...                 NaN
49             1992_WISCONSIN  ...                 NaN

[50 rows x 25 columns]

In [32]:
df.shape

(1715, 25)

In [94]:
df.isna().sum()*100/df.shape[0]

PRIMARY_KEY                      0.000000
STATE                            0.000000
YEAR                             0.000000
ENROLL                          28.629738
TOTAL_REVENUE                   25.655977
FEDERAL_REVENUE                 25.655977
STATE_REVENUE                   25.655977
LOCAL_REVENUE                   25.655977
TOTAL_EXPENDITURE               25.655977
INSTRUCTION_EXPENDITURE         25.655977
SUPPORT_SERVICES_EXPENDITURE    25.655977
OTHER_EXPENDITURE               28.629738
CAPITAL_OUTLAY_EXPENDITURE      25.655977
GRADES_PK_G                     10.087464
GRADES_KG_G                      4.839650
GRADES_4_G                       4.839650
GRADES_8_G                       4.839650
GRADES_12_G                      4.839650
GRADES_1_8_G                    40.524781
GRADES_9_12_G                   37.551020
GRADES_ALL_G                     4.839650
AVG_MATH_4_SCORE                67.055394
AVG_MATH_8_SCORE                64.897959
AVG_READING_4_SCORE             62

Common approaches involve filling in missing values in a variable with that variable's **mean**, **median**, or **mode** value.

But some variables contain blank values between 60-70%. Filling these variables with the values we have will not give an accurate result.

*   AVG_MATH_4_SCORE                67.055394
*   AVG_MATH_8_SCORE                64.897959
*   AVG_READING_4_SCORE             62.099125
*   AVG_READING_8_SCORE             67.230321





In [38]:
df['ENROLL'] = df.ENROLL.fillna(df.groupby('YEAR')['ENROLL'].transform('mean'))  
df['ENROLL'] 

0      NaN
1      NaN
2      NaN
3      NaN
4      NaN
        ..
1710   NaN
1711   NaN
1712   NaN
1713   NaN
1714   NaN
Name: ENROLL, Length: 1715, dtype: float64

There are still Nan values because there is no data for some years.
Can be filled with the average of data from recent years.

In [39]:
df['ENROLL'].iloc[90:100]

90     138939.0
91     631613.0
92     107050.0
93     844668.0
94    3535333.0
95     462323.0
96      97454.0
97    1032842.0
98     896197.0
99     317719.0
Name: ENROLL, dtype: float64

In [85]:
columns=['TOTAL_REVENUE','FEDERAL_REVENUE', 'STATE_REVENUE', 'LOCAL_REVENUE',
       'TOTAL_EXPENDITURE', 'INSTRUCTION_EXPENDITURE',
       'SUPPORT_SERVICES_EXPENDITURE', 'OTHER_EXPENDITURE',
       'CAPITAL_OUTLAY_EXPENDITURE', 'GRADES_PK_G', 'GRADES_KG_G',
       'GRADES_4_G', 'GRADES_8_G', 'GRADES_12_G', 'GRADES_1_8_G',
       'GRADES_9_12_G', 'GRADES_ALL_G', 'AVG_MATH_4_SCORE', 'AVG_MATH_8_SCORE',
       'AVG_READING_4_SCORE', 'AVG_READING_8_SCORE']

In [100]:
for column in columns:
    df[column] = df[column].fillna(df.groupby('YEAR')[column].transform('mean'))

In [101]:
df.isna().sum()*100/df.shape[0]

PRIMARY_KEY                      0.000000
STATE                            0.000000
YEAR                             0.000000
ENROLL                          28.629738
TOTAL_REVENUE                   24.139942
FEDERAL_REVENUE                 24.139942
STATE_REVENUE                   24.139942
LOCAL_REVENUE                   24.139942
TOTAL_EXPENDITURE               24.139942
INSTRUCTION_EXPENDITURE         24.139942
SUPPORT_SERVICES_EXPENDITURE    24.139942
OTHER_EXPENDITURE               27.230321
CAPITAL_OUTLAY_EXPENDITURE      24.139942
GRADES_PK_G                      3.090379
GRADES_KG_G                      3.090379
GRADES_4_G                       3.090379
GRADES_8_G                       3.090379
GRADES_12_G                      3.090379
GRADES_1_8_G                    39.358601
GRADES_9_12_G                   36.268222
GRADES_ALL_G                     3.090379
AVG_MATH_4_SCORE                59.825073
AVG_MATH_8_SCORE                59.825073
AVG_READING_4_SCORE             56

The basic idea in **interpolation** is to fill in missing values between two non-missing values using these non-missing values.

In [98]:
df3=df.interpolate(method='linear', limit_direction='forward', axis=0)
df3.isna().sum()*100/df.shape[0]

PRIMARY_KEY                      0.000000
STATE                            0.000000
YEAR                             0.000000
ENROLL                           2.973761
TOTAL_REVENUE                    0.000000
FEDERAL_REVENUE                  0.000000
STATE_REVENUE                    0.000000
LOCAL_REVENUE                    0.000000
TOTAL_EXPENDITURE                0.000000
INSTRUCTION_EXPENDITURE          0.000000
SUPPORT_SERVICES_EXPENDITURE     0.000000
OTHER_EXPENDITURE                2.973761
CAPITAL_OUTLAY_EXPENDITURE       0.000000
GRADES_PK_G                      0.000000
GRADES_KG_G                      0.000000
GRADES_4_G                       0.000000
GRADES_8_G                       0.000000
GRADES_12_G                      0.000000
GRADES_1_8_G                    14.868805
GRADES_9_12_G                   14.868805
GRADES_ALL_G                     0.000000
AVG_MATH_4_SCORE                 0.000000
AVG_MATH_8_SCORE                 0.000000
AVG_READING_4_SCORE              0

As a result, we can say that the best way to fill data in this data set, especially for score variables, is interpolation.